starfish/experiment/__init__.py

```diff
@@ -1,93 +1,172 @@
+        self._fovs = fovs
+        self._codebook = codebook
 
     @classmethod
-    def from_data(
-            cls,
-            image_stack: ImageStack,
-            aux_dict: Optional[Mapping[str, ImageStack]]=None
-    ) -> "Experiment":
-        """create a Stack from an already-loaded ImageStack
+    def from_json(cls, json_url: str) -> "Experiment":
+        """
+        Construct an `Experiment` from an experiment.json file format specifier
 
         Parameters
         ----------
-        image_stack : ImageStack
-            in-memory ImageStack
-        aux_dict : Optional[Mapping[str, ImageStack]]
-            a dictionary of ImageStacks, default None
+        json_url : str
+            file path or web link to an experiment.json file
 
         Returns
         -------
-        Stack :
-            a Stack object
+        Experiment :
+            Experiment object serving the requested experiment data
 
         """
-        stack = cls()
-        stack.image = image_stack
-        stack.auxiliary_images = aux_dict if aux_dict is not None else dict()
-        return stack
-
-    def read(self, in_json_path_or_url):
-        self.backend, name, self.baseurl = resolve_path_or_url(in_json_path_or_url)
-        with self.backend.read_contextmanager(name) as fh:
-            self.format_metadata = json.load(fh)
-
-        self.verify_version(self.format_metadata['version'])
-        self.image = ImageStack.from_url(self.format_metadata['hybridization_images'], self.baseurl)
-        for aux_key, aux_data in self.format_metadata['auxiliary_images'].items():
-            self.auxiliary_images[aux_key] = ImageStack.from_url(aux_data, self.baseurl)
+        backend, name, baseurl = resolve_path_or_url(json_url)
+        with backend.read_contextmanager(name) as fh:
+            experiment_document = json.load(fh)
+
+        cls.verify_version(experiment_document['version'])
+
+        _, codebook_name, codebook_baseurl = resolve_url(experiment_document['codebook'], baseurl)
+        codebook_absolute_url = pathjoin(codebook_baseurl, codebook_name)
+        codebook = Codebook.from_json(codebook_absolute_url)
+
+        hybridization_image: Union[Collection, TileSet] = Reader.parse_doc(
+            experiment_document['hybridization_images'], baseurl)
+        auxiliary_images: MutableMapping[str, Union[Collection, TileSet]] = dict()
+        for aux_image_type, aux_image_url in experiment_document['auxiliary_images'].items():
+            auxiliary_images[aux_image_type] = Reader.parse_doc(aux_image_url, baseurl)
+
+        if isinstance(hybridization_image, TileSet):
+            # make sure that all the aux images are also TileSets
+            for aux_image_tileset in auxiliary_images.values():
+                assert isinstance(aux_image_tileset, TileSet)
+
+            fov = FieldOfView(
+                None,
+                primary_image_tileset=hybridization_image,
```

Comment:
codebase is starting to get weird about primary_image v.s hybridization_images. since the PR in spaceTx fmt to rename hybridization_images->primary_images, i think all new code should start using this language?

Reply:
Sure, I will make this change.